Added guide about proxy exporters/blackbox exporter (#1209)
prometheus/docs: content/docs/guides/multi-target-exporter.md

* describe the multi-target exporter pattern and why it is used,
* run the [blackbox](https://github.com/prometheus/blackbox_exporter) exporter as an example of the pattern,
* configure a custom query module for the blackbox exporter,
* let the blackbox exporter run basic metric queries against the Prometheus [website](https://prometheus.io),
* examine a popular pattern of configuring Prometheus to scrape exporters using relabeling.

## The multi-target exporter pattern?

By multi-target [exporter](/docs/instrumenting/exporters/) pattern we refer to a specific design, in which:

* the exporter will get the target’s metrics via a network protocol.
* the exporter does not have to run on the machine the metrics are taken from.
* the exporter gets the targets and a query config string as parameters of Prometheus’ GET request.
* the exporter subsequently starts the scrape after getting Prometheus’ GET requests and once it is done with scraping.
* the exporter can query multiple targets.

This pattern is only used for certain exporters, namely the [blackbox](https://github.com/prometheus/blackbox_exporter) and the [SNMP exporter](https://github.com/prometheus/snmp_exporter).

> The reason is that we either can’t run an exporter on the targets, e.g. network gear speaking SNMP, or that we are explicitly interested in the distance, e.g. latency and reachability of a website from a specific point outside of our network.

Reviewer: Mention blackbox for the second type? You're using it for examples below so that gives some mental reference.

Author: ```suggestion
The reason is that we either can’t run an exporter on the targets, e.g. network gear speaking SNMP, or that we are explicitly interested in the distance, e.g. latency and reachability of a website from a specific point outside of our network, a common use case for the [blackbox](https://github.com/prometheus/blackbox_exporter) exporter.
```

## Running multi-target exporters

Multi-target exporters are flexible regarding their environment and can be run in many ways. As regular programs, in containers, as background services, on baremetal, virtual machines. Because they are queried and do query over network they do need appropriate open ports. Otherwise they are frugal.

Now let’s try it out for yourself!

Use [Docker](https://www.docker.com/) to start a blackbox exporter container by running this in a terminal. Depending on your system configuration you might need to prepend the command with a `sudo`:

```bash
docker run -p 9115:9115 prom/blackbox-exporter
```

You should see a few log lines and if everything went well the last one should report `msg="Listening on address"` as seen here:

```
level=info ts=2018-10-17T15:41:35.4997596Z caller=main.go:324 msg="Listening on address" address=:9115
```

## Basic querying of multi-target exporters

There are two ways of querying:

1. Querying the exporter itself. It has its own metrics, usually available at `/metrics`.

> 1. Querying the exporter to scrape another target. Usually available at a "descriptive" endpoint, e.g. `/probe`.

Reviewer: Maybe state that this is what people usually want but that the former is better for demonstration?

Author: ```suggestion
1. Querying the exporter to scrape another target. Usually available at a "descriptive" endpoint, e.g. `/probe`. This is likely what you are primarily interested in, when using multi-target exporters.
```

  metrics_path: /metrics
  static_configs:
    - targets:
      - localhost:9115   # for Windows and macOS replace with - host.docker.internal:9115

- job_name: 'blackbox'
  scrape_interval: 5s
  metrics_path: /probe
  params:
    module: [http_2xx]
    target: [prometheus.io]
  static_configs:
    - targets:
      - localhost:9115   # for Windows and macOS replace with - host.docker.internal:9115
```

After saving the config file switch to the terminal with your Prometheus docker container and stop it by pressing `ctrl+C` and start it again to reload the configuration by using the existing [command](#run-prometheus).

The terminal should return the message `"Server is ready to receive web requests."` and after a few seconds you should start to see colourful graphs in [your Prometheus](http://localhost:9090/graph?g0.range_input=5m&g0.stacked=0&g0.expr=probe_http_duration_seconds&g0.tab=0).

This works, but it has a few disadvantages:

1. The actual targets are up in the param config, which is very unusual and hard to understand later.
1. The `instance` label has the value of the blackbox exporter’s address which is technically true, but not what we are interested in.

> 1. We can’t see which URL we probed. This is unpractical and will also mix up different metrics into one if we probe several URLs.

Reviewer: Maybe "target URL"?

To fix this, we will use [relabeling](/docs/prometheus/latest/configuration/configuration/#<relabel_config>).

> Relabeling is very powerful for advanced usecases because behind the scenes many things in Prometheus are configured with internal labels.

Reviewer: This sentence is somewhat complicated and doesn't really tell the reader much.

> The details are worthy of a [talk](https://www.youtube.com/watch?v=b5-SvvZ7AwI) on its own. For now is suffices if you understand this:

Reviewer: Maybe rephrase that the the details are complicated, outside of scope, you will limit yourself to what's needed here and then refer to the talk as further info?

Here is the config you will use to do that. Don’t worry if this is a bit much at once, we will go through it step by step:

```yaml
scrape_configs:

- job_name: blackbox-exporter-metrics
  scrape_interval: 5s
  metrics_path: /metrics
  static_configs:
    - targets:
      - localhost:9115   # for Windows and macOS replace with - host.docker.internal:9115

- job_name: 'blackbox'
  scrape_interval: 5s
  metrics_path: /probe
  params:
    module: [http_2xx]
  static_configs:
      - targets:
        - http://prometheus.io    # Target to probe with http.
        - https://prometheus.io   # Target to probe with https.
        - http://example.com:8080 # Target to probe with http on port 8080.
  relabel_configs:
    - source_labels: [__address__]
      target_label: __param_target
    - source_labels: [__param_target]
      target_label: instance
    - target_label: __address__
      replacement: localhost:9115  # The blackbox exporter’s real hostname:port. For Windows and macOS replace with - host.docker.internal:9115
```

> So what is new compared to the [last config](#prometheus-config)?

Reviewer: I like the below, but maybe split it in "results" and "what are the steps to get there ", this again subsplit for each of the three sections under `relabel_configs`?

Author: Good point.
I tried to change this using the suggestion feature, but github is confused by nested markdown code blocks within suggestions. I will change it directly. Sorry for the inconvenience.

Reviewer: Looking good.

 1. `params` does not include `target` anymore.
 1. We add the actual targets under `static configs:` `targets`. We also use several because we can do that now.
 1. `relabel_configs` contains the new relabeling rules.
 1. We take the values from the label `__address__` (which contain the values from `targets`) and write them to a new label `__param_target` which will add a parameter `target` to the Prometheus scrape requests.
 1. We take the values from the label `__param_target` and create a label instance with the values.
 1. We write the value `localhost:9115` to the the label `__address__`. This will be used as the hostname and port for the Prometheus scrape requests.

This way we can have the actual targets there, get them as `instance` label values while letting Prometheus make a request against the blackbox exporter.

Often people combine these with a specific service discovery. Check out the [configuration documentation](/docs/prometheus/latest/configuration/configuration) for more information. Using them is no problem, as these write into the `__address__` label just like `targets` defined under `static_configs`.

> That is it. Restart the Prometheus docker container and look at your [metrics](http://localhost:9090/graph?g0.range_input=5m&g0.stacked=0&g0.expr=probe_http_duration_seconds&g0.tab=0).

Reviewer: Maybe a closing sentence about what differences to look for, or how to extend the time range of the graph in case they took longer to do it?

> That is it. Restart the Prometheus docker container and look at your [metrics](http://localhost:9090/graph?g0.range_input=5m&g0.stacked=0&g0.expr=probe_http_duration_seconds&g0.tab=0).

Author: ```suggestion
That is it. Restart the Prometheus docker container and look at your [metrics](http://localhost:9090/graph?g0.range_input=30m&g0.stacked=0&g0.expr=probe_http_duration_seconds&g0.tab=0). Pay attention that you selected the period of time when the metrics were actually collected.
```

---
title: Understanding and using the multi-target exporter pattern
---

# Understanding and using the multi-target exporter pattern

This guide will introduce you to the multi-target exporter pattern. To achieve this we will:

* describe the multi-target exporter pattern and why it is used,
* run the [blackbox](https://github.com/prometheus/blackbox_exporter) exporter as an example of the pattern,
* configure a custom query module for the blackbox exporter,
* let the blackbox exporter run basic metric queries against the Prometheus [website](https://prometheus.io),
* examine a popular pattern of configuring Prometheus to scrape exporters using relabeling.

## The multi-target exporter pattern?

By multi-target [exporter](/docs/instrumenting/exporters/) pattern we refer to a specific design, in which:

* the exporter will get the target’s metrics via a network protocol.

> * the exporter does not have to run on the machine the metrics are taken from.

Reviewer 2: (almost) no exporter has to run on the same machine, so this is a bit misleading.

Author: Some do. Especially the node-exporter, which is among the first most people come in contact with. This is a guide aimed at people relatively new to Prometheus and I want to make it easy to grasp the concepts while saying the truth.

Reviewer 2: This point is misleading though. That you can run e.g. the haproxy exporter on a different machine doesn't make it multi-target.

Author: All of the points have to be true. One is not sufficient. To me that was clear by "in which". But we can change that. Any ideas?

Reviewer 2: It's a bit ambiguous as written. "does not usually" maybe?

Author: I am a bit confused. Where exactly would you put the "does not usually"?

Reviewer 2: the exporter does not usually run on the machine

* the exporter gets the targets and a query config string as parameters of Prometheus’ GET request.
* the exporter subsequently starts the scrape after getting Prometheus’ GET requests and once it is done with scraping.
* the exporter can query multiple targets.

> This pattern is only used for certain exporters, namely the [blackbox](https://github.com/prometheus/blackbox_exporter) and the [SNMP exporter](https://github.com/prometheus/snmp_exporter).

Reviewer 2: s/namely/such as/

These aren't the only ones.

Author: Good catch. Namely is a false friend between German and English I was not aware of. Thank you for pointing it out.

The reason is that we either can’t run an exporter on the targets, e.g. network gear speaking SNMP, or that we are explicitly interested in the distance, e.g. latency and reachability of a website from a specific point outside of our network, a common use case for the [blackbox](https://github.com/prometheus/blackbox_exporter) exporter.

## Running multi-target exporters

> Multi-target exporters are flexible regarding their environment and can be run in many ways. As regular programs, in containers, as background services, on baremetal, virtual machines. Because they are queried and do query over network they do need appropriate open ports. Otherwise they are frugal.

Reviewer 2: on virtual machines.

I'm not sure this sentence structure will work for non-native speakers.

Author: I tested the wording and the guide as a whole with non-native speakers, hence I am optimistic.

Now we need to tell the blackbox exporter to use our freshly changed file. We can do that with the flag `--config.file="blackbox.yml"`. But because we are using Docker, we first must make this file [available](https://docs.docker.com/storage/bind-mounts/) inside the container using the `--mount` command.  

NOTE: If you are using macOS you first need to allow the Docker daemon to access the directory in which your `blackbox.yml` is. You can do that by clicking on the little Docker whale in menu bar and then on `Preferences`->`File Sharing`->`+`. Afterwards press `Apply & Restart`.

First you stop the old container by changing into its terminal and press `ctrl+c`.
Make sure you are in the directory containing your `blackbox.yml`.
Then you run this command. It is long, but we will explain it:

<a name="run-exporter"></a>

```bash
docker \
  run -p 9115:9115 \
  --mount type=bind,source="$(pwd)"/blackbox.yml,target=/blackbox.yml,readonly \
  prom/blackbox-exporter \
  --config.file="/blackbox.yml"
```

With this command, you told `docker` to:

1. `run` a container with the port `9115` outside the container mapped to the port `9115` inside of the container.

> 1. `mount` from your current directory (`$(pwd)` stands for *p*rint *w*orking *d*irectory) the file `blackbox.yml` into `/blackbox.yml` in `readonly` mode.

Author: Does it? It looks fine in the GitHub file preview.
But my heart is not attached to the emphasis of the initials.
If you think it is better, I can take them away.

> 1. `mount` from your current directory (`$(pwd)` stands for *p*rint *w*orking *d*irectory) the file `blackbox.yml` into `/blackbox.yml` in `readonly` mode.

Reviewer 3: In the diff view everything afterwards is italic. To me that shows that it's at least ambiguous, Markdown being the underspecified mess that it is. I'd say remove it to avoid issues

> 1. `mount` from your current directory (`$(pwd)` stands for *p*rint *w*orking *d*irectory) the file `blackbox.yml` into `/blackbox.yml` in `readonly` mode.

Author: Well, sad, but okay. I’ll remove the *s.

NOTE: If you use Docker for Mac or Docker for Windows, you can’t use `localhost:9115` in the last line, but must use `host.docker.internal:9115`. This has to do with the virtual machines used to implement Docker on those operating systems. You should not use this in production.

`prometheus.yml` for Linux:

```yaml
scrape_configs:

> - job_name: blackbox-exporter-metrics

Reviewer 2: This is a bit verbose, and metrics is redundant here. Just blackbox will do.

Author: I get why you say that, but this is to clarify that it is about the metrics of the exporter itself.
It is aimed at people who want to learn about the concept and named like that for pedagogical reasons.

Reviewer 2: It's redundant though, and I consider it an anti-pattern. This will be blindly copy&pasted, so we should follow best practices.

Author: How about `blackbox-exporter-itself`?

Reviewer 2: That's still a bit redundant, I'd personally use blackbox, but could live with blackbox_exporter

Author: Redundancy is a feature in this context.

Reviewer 2: It's a bug, as it's extra typing/reading. We don't need every job name ending with prometheus-telemetry-metrics-job (I've seen much worse in practice). Job names should be succinct.

Author: Yes, in a normal context it is absolutely a bug. I don’t disagree with you. And I share your frustration about redundant job names in production systems.
But this is an exercise to help people form their mind. We are trying to solve a different problem than usually. Everything is only true within a context and when it changes things can become tricky.
Maybe we can find a solution that satisfies both requirements at once? Something pedagogical and not to verbose.

We have two jobs:
* One about the metrics of the exporter itself. That must be clear to people who are new.
* One about the targets we tell the exporter to test for us. That must also be clear for new people and not implying magic. 

At the same time we want people not to form the bad habit of really long redundant job names.

How about we use a comment? It is a bit worse for understanding and less elegant, but the risk that people adopt bad habits is much lower.

```yaml
 - job_name: blackbox-exporter # to get metrics about the exporter itself 
[…]
 - job_name: blackbox # to get metrics from targets we want the exporter to query for us
```

Reviewer 4: I think you underestimate how we just all copy and past examples from the internet. I am on Brian's side on this one.

Author: Good point. I’ll go with the comments solution.

Reviewer 2: We can use comments reasonably liberally, the type of probe should be in the job name though as you can't put all your different probe types in one job.

>   scrape_interval: 5s

Reviewer 2: Per scrape-config intervals are generally an anti-pattern, do this up in global.

Author: Will do.

  metrics_path: /metrics
  static_configs:
    - targets:
      - localhost:9115
```

`prometheus.yml` for macOS and Windows:

```yaml
scrape_configs:

- job_name: blackbox-exporter-metrics
  scrape_interval: 5s
  metrics_path: /metrics
  static_configs:
    - targets:
      - host.docker.internal:9115
```

Now run a Prometheus container and tell it to mount our config file from above. Because of the way networking on the host is addressable from the container you need to use a slightly different command on Linux than on MacOS and Windows.:

<a name=run-prometheus></a>

Run Prometheus on Linux (don’t use `--network="host"` in production):

```bash
docker \
  run --network="host"\

>   --mount type=bind,source="$(pwd)"/prometheus.yml,target=/prometheus.yml,readonly \

Reviewer 2: Did you mean $PWD?

Author: No, I didn’t.
Please try to avoid passive-aggressive language. It is off-putting and not good for fostering collaboration.

Reviewer 2: Hmm, I didn't realise that this works.

It was an honest question.

Author: I am sorry Brian. I read it like a smug "please use this instead". I am very glad you did not mean it like that.
All of this works, I tested it under both, Linux and Mac OS myself and let a new to prometheus engineer try the whole guide to see how it does from a pedagogical perspective as well.

Reviewer 2: Ah no, I'd actually mis-read it as {} rather than () and thus thought that this was a error.

  --mount type=bind,source="$(pwd)"/prometheus.yml,target=/prometheus.yml,readonly \
  prom/prometheus \
  --config.file="/prometheus.yml"
```

This command works similarly to [running the blackbox exporter using a config file](#run-exporter).

If everything worked, you should be able to go to [localhost:9090/targets](http://localhost:9090/targets) and see under `blackbox-exporter-metrics` an endpoint with the state `UP` in green. If you get a red `DOWN` make sure that the blackbox exporter you started [above](run-exporter) is still running. If you see nothing or a yellow `UNKNOWN` you are really fast and need to give it a few more seconds before reloading your browser’s tab.

To tell Prometheus to query `"localhost:9115/probe?target=prometheus.io&module=http_2xx"` you add another scrape job where you set the `metrics_path` to `/probe` and the parameters under `params:` in the Prometheus config file `prometheus.yml`:

<a name="prometheus-config"></a>

```yaml
scrape_configs:

- job_name: blackbox-exporter-metrics
  scrape_interval: 5s
  metrics_path: /metrics
  static_configs:
    - targets:
      - localhost:9115   # for Windows and macOS replace with - host.docker.internal:9115

> - job_name: 'blackbox'

Reviewer 2: This isn't monitoring blackbox so this is misleading. blackbox_http or similar would be clearer.

Author: Good Point.

Reviewer 2: It's still important in terms of label taxonomy, it means something to humans and that's importnat.

Author: Yes. Exactly because it does mean something for humans I left it aside.

Reviewer 2: But it's not sufficiently distinguishing, you wouldn't be mixing http and tcp probes under the same job. This name to me could also realistically mean the exporter itself.

Author: I propose we discuss this further up at the other job name. Those two issues are connected anyway.

Author: Resolving that just to make the view clearer. The issue will be discussed above.

> We can see that the probe was successful and get many useful metrics, like latency by phase, status code, ssl status or cert expiry in [epoch](https://en.wikipedia.org/wiki/Unix_time).  

Author: Sure, I will change it as you said.

    - target_label: __address__
      replacement: localhost:9115  # The blackbox exporter’s real hostname:port. For Windows and macOS replace with - host.docker.internal:9115
```

So what is new compared to the [last config](#prometheus-config)?

`params` does not include `target` anymore. Instead we add the actual targets under `static configs:` `targets`. We also use several because we can do that now:
```yaml
  params:
    module: [http_2xx]
  static_configs:
      - targets:
        - http://prometheus.io    # Target to probe with http.
        - https://prometheus.io   # Target to probe with https.
        - http://example.com:8080 # Target to probe with http on port 8080.
```

`relabel_configs` contains the new relabeling rules:
```yaml
  relabel_configs:
    - source_labels: [__address__]
      target_label: __param_target
    - source_labels: [__param_target]
      target_label: instance
    - target_label: __address__
      replacement: localhost:9115  # The blackbox exporter’s real hostname:port. For Windows and macOS replace with - host.docker.internal:9115
```

Before applying the relabeling rules, the URI of a request Prometheus would make would look like this:

> `"http://prometheus.io/probe?module=http_2xx"`. After relabeling it will look like this `"localhost:9115/probe?target=http://prometheus.io&module=http_2xx"`.

Reviewer 2: The leading http should still be there.

Author: True. Thanks for pointing that out.

```yaml
  relabel_configs:
    - target_label: __address__
      replacement: localhost:9115  # The blackbox exporter’s real hostname:port. For Windows and macOS replace with - host.docker.internal:9115
```

Our request is now `"localhost:9115/probe?target=http://prometheus.io&module=http_2xx"`. This way we can have the actual targets there, get them as `instance` label values while letting Prometheus make a request against the blackbox exporter.

Often people combine these with a specific service discovery. Check out the [configuration documentation](/docs/prometheus/latest/configuration/configuration) for more information. Using them is no problem, as these write into the `__address__` label just like `targets` defined under `static_configs`.

That is it. Restart the Prometheus docker container and look at your [metrics](http://localhost:9090/graph?g0.range_input=30m&g0.stacked=0&g0.expr=probe_http_duration_seconds&g0.tab=0). Pay attention that you selected the period of time when the metrics were actually collected.

# Summary

> In this guide, you learned how the multi-target exporter pattern works and how to run a blackbox exporter with a customised module and to configure Prometheus using relabeling to scrape metrics with prober labels.

Reviewer 2: This is a bit run on, replace the first and with a comma.

Author: Will do.

* describe the multi-target exporter pattern and why it is used,
* run the [blackbox](https://github.com/prometheus/blackbox_exporter) exporter as an example of the pattern,
* configure a custom query module for the blackbox exporter,
* let the blackbox exporter run basic metric queries against the Prometheus [website](https://prometheus.io),
* examine a popular pattern of configuring Prometheus to scrape exporters using relabeling.

## The multi-target exporter pattern?

By multi-target [exporter](/docs/instrumenting/exporters/) pattern we refer to a specific design, in which:

* the exporter will get the target’s metrics via a network protocol.
* the exporter does not have to run on the machine the metrics are taken from.
* the exporter gets the targets and a query config string as parameters of Prometheus’ GET request.
* the exporter subsequently starts the scrape after getting Prometheus’ GET requests and once it is done with scraping.
* the exporter can query multiple targets.

This pattern is only used for certain exporters, namely the [blackbox](https://github.com/prometheus/blackbox_exporter) and the [SNMP exporter](https://github.com/prometheus/snmp_exporter).

> The reason is that we either can’t run the exporter on the targets, e.g. network gear speaking SNMP, or that we are explicitly interested in the distance, e.g. latency and reachability of a website from a specific point outside of our network.

Reviewer 2: an exporter

## Running multi-target exporters

Multi-target exporters are flexible regarding their environment and can be run in many ways. As regular programs, in containers, as background services, on baremetal, virtual machines. Because they are queried and do query over network they do need appropriate open ports. Otherwise they are frugal.

Now let’s try it out for yourself!

Use [Docker](https://www.docker.com/) to start a blackbox exporter container running by running this in a terminal:

```bash

> sudo docker run -p 9115:9115 prom/blackbox-exporter

Reviewer 2: Why the sudo?

Author: Because this docker command does not work without using sudo.

Reviewer 2: I don't understand why this should require sudo, this sounds like something odd with your setup rather than a general thing about docker.

Author: I tested this on a very vanilla Ubuntu Linux like many potential readers of the guide would use and it needs sudo. What more can I tell you about it?

Reviewer 2: The command works fine for me without sudo, and I'm also on vanilla Ubuntu. I'm guessing your user isn't a member of the docker group?

In generally let's not recommend use of sudo when it's not 100% required.

Reviewer 3: I haven't tried it yet but docker on Mac / Windows is strongly tied to the _user_ so I expect `sudo docker` tp either blow up spectacularly or misbehave subtly. let's give the command without it, and mention in a separate sentence that in some setups you may need sudo?

Author: Okay, that sounds like a reasonable approach between best practices and raising the bar to entry. I’ll add it.

```

You should see a few log lines and if everything went well the last one should report `msg="Listening on address"` as seen here:

```
level=info ts=2018-10-17T15:41:35.4997596Z caller=main.go:324 msg="Listening on address" address=:9115
```

## Basic querying of multi-target exporters

There are two ways of querying:

1. Querying the exporter itself. It has its own metrics, usually available at `/metrics`.
1. Querying the exporter to scrape another target. Usually available at a "descriptive" endpoint, e.g. `/probe`.

> You can manually try the first query type with curl in another terminal. If you don’t have curl, you can also copy the part between the `'` into a browser’s address bar:

Reviewer 2: I think it's reasonably to presume that everyone has curl, otherwise use an `<a>` to link there directly.

Author: You are right, most people do have curl, but it is not preinstalled on many popular distributions.
A link is a very good idea, I will change that.

> <a name="query-exporter"></a>

Reviewer 2: Is this meant to be here?

Author: Yes, I link to that anchor later in the guide to make it more comfortable to compare two queries for interested users.

```bash
curl 'localhost:9115/metrics'
```

The response should be something like this:

```

> # HELP blackbox_exporter_build_info A metric with a constant '1' value labeled by version, revision, branch, and goversion from which blackbox_exporter was built.

Reviewer 2: I'd trim this down, it's largely not relevant to the goal of the page

Author: Sure.

Like this:
```
TEXT text text text
TEXT text text text
some_metric 0

…

HELP text text text
TEXT text text text
another_metric 1
```

or this:

```
TEXT text text text
TEXT text text text
some_metric 0

[…]

HELP text text text
TEXT text text text
another_metric 1
```

Reviewer 2: Whatever you think is best

# TYPE probe_http_redirects gauge
probe_http_redirects 1
# HELP probe_http_ssl Indicates if SSL was used for the final redirect
# TYPE probe_http_ssl gauge
probe_http_ssl 1
# HELP probe_http_status_code Response HTTP status code
# TYPE probe_http_status_code gauge
probe_http_status_code 200
# HELP probe_http_uncompressed_body_length Length of uncompressed response body
# TYPE probe_http_uncompressed_body_length gauge
probe_http_uncompressed_body_length 14516
# HELP probe_http_version Returns the version of HTTP of the probe response
# TYPE probe_http_version gauge
probe_http_version 1.1
# HELP probe_ip_protocol Specifies whether probe ip protocol is IP4 or IP6
# TYPE probe_ip_protocol gauge
probe_ip_protocol 4
# HELP probe_ssl_earliest_cert_expiry Returns earliest SSL cert expiry in unixtime
# TYPE probe_ssl_earliest_cert_expiry gauge
probe_ssl_earliest_cert_expiry 1.581897599e+09
# HELP probe_success Displays whether or not the probe was a success
# TYPE probe_success gauge
probe_success 1
# HELP probe_tls_version_info Contains the TLS version used
# TYPE probe_tls_version_info gauge
probe_tls_version_info{version="TLS 1.3"} 1
```

> You can see that the probe was successful and get many useful metrics, like latency by phase, status code, ssl status or cert expiry in [epoch](https://en.wikipedia.org/wiki/Unix_time).  

Reviewer 2: I wouldn't refer to this as just "epoch", which epoch? What unit?
"Unix time" would be the usual term.
